gprofiler/profilers/java.py

```diff
@@ -259,11 +284,107 @@ def make_application_metadata(self, process: Process) -> Dict[str, Any]:
+            return False
+
+        return True
+
+    def get_jvm_flags_raw(self, process: Process) -> Optional[List[JvmFlag]]:
+        try:
+            output = self.jattach_jcmd_runner.run(process, "VM.flags -all")
+        except CalledProcessError as e:
+            logging.warning(f"Couldn't get jvm flags for process {process.pid}: {e.stderr}")
+            return None
+
+        vm_flags = []
+        # The output of VM.flags -all format on jdk 8:
+        # bool UseCompressedClassPointers               := true                                {lp64_product}
+        # flag_type flag_name                           := flag_value                          {flag_kind}
+        # ":=" indicates non default origin for the flag, while "=" indicates default origin
+
+        # The output of VM.flags -all format on jdk 9+:
+        # bool OptoScheduling                           = false                               {C2 pd product} {default}
+        # flag_type flag_name                           = flag_value                          {flag_kind} {flag_origin}
+
+        # flag_kind is space separated list of kinds, e.g. "C2 pd product"
+
+        # possible flag kinds: "product", "manageable", "diagnostic", "experimental", "notproduct", "develop",
+        # "lp64_product", "rw", "pd", "JVMCI", "C1", "C2", "ARCH"
+
+        # possible flag types: "bool", "int", "uint", "intx", "uintx", "uint64_t", "size_t", "double", "ccstr",
+        # "ccstrlist"
+
+        # possible flag origins: default, non-default, command line, environment, config file, management, ergonomic,
+        # attach, internal, jimage
+
+        vm_flags_pattern = re.compile(
+            r"(?P<flag_type>\S+)\s+(?P<flag_name>\S+)\s+(?P<flag_is_non_default_origin_only_jdk_8>:)?= (?P<flag_value>\S*)\s+{(?P<flag_kind>.+?)}(?:\s*{(?P<flag_origin_jdk_9>.*)})*"  # noqa: E501
+        )
+
+        for line in output.splitlines():
+            match = vm_flags_pattern.search(line)
+            if match:
+                # get the flag origin if jvm 9+, otherwise get is the flag from non default origin as described above
+                flag_is_non_default_origin_only_jdk_8 = match.group("flag_is_non_default_origin_only_jdk_8") == ":"
+                flag_origin_jdk_9 = match.group("flag_origin_jdk_9")
+
+                # split the list of space separated flag_kinds as described above
+                flag_kind = match.group("flag_kind").split()
+
+                flag_name = match.group("flag_name")
+                flag_value = match.group("flag_value")
+                flag_type = match.group("flag_type")
+
+                if flag_is_non_default_origin_only_jdk_8:
+                    flag_origin = "non-default"
+                elif flag_origin_jdk_9:
+                    flag_origin = flag_origin_jdk_9
+                else:
+                    flag_origin = "default"
+
+                if flag_type and flag_name and flag_value and flag_kind:
```

Comment:
Isn't there a way to make the `search` method return a match object only when all non-optional groups were matched? I hoped this is the default behavior.

Reply:
none that I know of

Reply:
Are you sure about it?
``` python
In [14]: re.search(r"(?P<flag_type>\S+)\s+(?P<flag_name>\S+)\s+(?P<flag_is_non_default_origin_only_jdk_8>:)?= (?P<flag_value>\S*)\s+{(?P<flag_kind>.+?)}(?:\s*{(?P<flag_origin_jdk_9>.*)})?", "t n = v {k} {o}") is None
Out[14]: False

In [15]: re.search(r"(?P<flag_type>\S+)\s+(?P<flag_name>\S+)\s+(?P<flag_is_non_default_origin_only_jdk_8>:)?= (?P<flag_value>\S*)\s+{(?P<flag_kind>.+?)}(?:\s*{(?P<flag_origin_jdk_9>.*)})?", "t n = v {k}") is None
Out[15]: False

In [16]: re.search(r"(?P<flag_type>\S+)\s+(?P<flag_name>\S+)\s+(?P<flag_is_non_default_origin_only_jdk_8>:)?= (?P<flag_value>\S*)\s+{(?P<flag_kind>.+?)}(?:\s*{(?P<flag_origin_jdk_9>.*)})?", "t n =  {k}") is None
Out[16]: False

In [17]: re.search(r"(?P<flag_type>\S+)\s+(?P<flag_name>\S+)\s+(?P<flag_is_non_default_origin_only_jdk_8>:)?= (?P<flag_value>\S*)\s+{(?P<flag_kind>.+?)}(?:\s*{(?P<flag_origin_jdk_9>.*)})?", "t n =  ") is None
Out[17]: True
```
The issue is only with `t n =  {k}` where the value is matched to `""`, but fixing `(?P<flag_value>\S*)` to `(?P<flag_value>\S+)` solves that:
``` python
In [22]: re.search(r"(?P<flag_type>\S+)\s+(?P<flag_name>\S+)\s+(?P<flag_is_non_default_origin_only_jdk_8>:)?= (?P<flag_value>\S+)\s+{(?P<flag_kind>.+?)}(?:\s*{(?P<flag_origin_jdk_9>.*)})?", "t n = v {k} {o}") is None
Out[22]: False

In [23]: re.search(r"(?P<flag_type>\S+)\s+(?P<flag_name>\S+)\s+(?P<flag_is_non_default_origin_only_jdk_8>:)?= (?P<flag_value>\S+)\s+{(?P<flag_kind>.+?)}(?:\s*{(?P<flag_origin_jdk_9>.*)})?", "t n = v {k}") is None
Out[23]: False

In [24]: re.search(r"(?P<flag_type>\S+)\s+(?P<flag_name>\S+)\s+(?P<flag_is_non_default_origin_only_jdk_8>:)?= (?P<flag_value>\S+)\s+{(?P<flag_kind>.+?)}(?:\s*{(?P<flag_origin_jdk_9>.*)})?", "t n = {k}") is None
Out[24]: True

In [25]: re.search(r"(?P<flag_type>\S+)\s+(?P<flag_name>\S+)\s+(?P<flag_is_non_default_origin_only_jdk_8>:)?= (?P<flag_value>\S+)\s+{(?P<flag_kind>.+?)}(?:\s*{(?P<flag_origin_jdk_9>.*)})?", "t n = v") is None
Out[25]: True
```

Reply:
"" is indeed a legal value

Reply:
Discussion continues in https://github.com/Granulate/granulate-utils/pull/122.